### PR TITLE
[BUG] Trying to fix NaN values produced by summarize.py

### LINKED ISSUE #5781: [BUG] WindowSummarizer unexpected NaN
**Describe the bug**

With Sktime 0.25.0 the WindowSummarizer returns unexpected missing values.

**To Reproduce**

```python
import numpy as np
import pandas as pd
from sktime.transformations.series.summarize import WindowSummarizer

df = pd.DataFrame(np.arange(10))
transformer = WindowSummarizer(lag_feature={"mean": [[-1, 1], [-1, 2]]}, n_jobs=1)
df_transformed = transformer.fit_transform(df)

print(df_transformed)
```

**Expected behavior**

value 0.5 in 1st row 2nd column rather than NaN

**Additional context**

**Versions**
<details>

System:
    python: 3.10.12 | packaged by Anaconda, Inc. | (main, Jul  5 2023, 19:01:18) [MSC v.1916 64 bit (AMD64)]
executable: [c:\Users\ramon.bussing\AppData\Local\anaconda3\envs\enerfore-dev\python.exe](file:///C:/Users/Ram0nB/AppData/Local/anaconda3/envs/enerfore-dev/python.exe)
   machine: Windows-10-10.0.19044-SP0

Python dependencies:
          pip: 23.2.1
       sktime: 0.25.0
      sklearn: 1.3.0
       skbase: 0.6.1
        numpy: 1.22.4
        scipy: 1.8.1
       pandas: 2.1.1
   matplotlib: None
       joblib: 1.3.2
        numba: None
  statsmodels: 0.14.0
     pmdarima: None
statsforecast: None
      tsfresh: None
      tslearn: None
        torch: None
   tensorflow: None
tensorflow_probability: None

</details>

### PR DESCRIPTION
**Pull request by ShreeshaM07**
#### Reference Issues/PRs

Resolves #5781 

#### What does this implement/fix? Explain your changes.

This tries to fix the NaN values produced in the given code snippet in the issue and prints 0.5 as 1st row 2nd column . However this does not work for some examples like in `./sktime/examples/01c_forecasting_hierarchical_global.ipynb`. It gives errors like 
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[28], line 17
      7 kwargs = {
      8     "lag_feature": {
      9         "lag": [1],
   (...)
     12     }
     13 }
     16 transformer = WindowSummarizer(**kwargs)
---> 17 y_transformed = transformer.fit_transform(y_train)
     19 y_transformed.head(10)

File [~/sktime_shreesha/sktime/sktime/transformations/base.py:681](https://file+.vscode-resource.vscode-cdn.net/home/shreesha/sktime_shreesha/sktime/examples/~/sktime_shreesha/sktime/sktime/transformations/base.py:681), in BaseTransformer.fit_transform(self, X, y)
    616 """Fit to data, then transform it.
    617 
    618 Fits the transformer to X and y and returns a transformed version of X.
   (...)
    677         Example: i-th instance of the output is the i-th window running over `X`
    678 """
    679 # Non-optimized default implementation; override when a better
    680 # method is possible for a given algorithm.
--> 681 return self.fit(X, y).transform(X, y)

File [~/sktime_shreesha/sktime/sktime/transformations/base.py:595](https://file+.vscode-resource.vscode-cdn.net/home/shreesha/sktime_shreesha/sktime/examples/~/sktime_shreesha/sktime/sktime/transformations/base.py:595), in BaseTransformer.transform(self, X, y)
...
   2469         yield recode_for_categories(
   2470             self.codes[i], self.levels[i], new_levels[i], copy=copy
   2471         )

AssertionError: Length of new_levels (3) must be <= self.nlevels (2)
```
upon running 
```
import pandas as pd

from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import ForecastingPipeline
from sktime.transformations.series.summarize import WindowSummarizer

kwargs = {
    "lag_feature": {
        "lag": [1],
        "mean": [[1, 3], [3, 6]],
        "std": [[1, 4]],
    }
}


transformer = WindowSummarizer(**kwargs)
y_transformed = transformer.fit_transform(y_train)

y_transformed.head(10)
```
and on running this code block
```
import numpy as np


def count_gt130(x):
    """Count how many observations lie above threshold 130."""
    return np.sum((x > 700)[::-1])


kwargs = {
    "lag_feature": {
        "lag": [1],
        count_gt130: [[2, 3]],
        "std": [[1, 4]],
    }
}

transformer = WindowSummarizer(**kwargs)
y_transformed = transformer.fit_transform(y_train)

y_transformed.head(10)
```
it gives the error
```
Traceback (most recent call last):
  File "/home/shreesha/anaconda3/envs/sktime-dev/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py", line 463, in _process_worker
    r = call_item()
        ^^^^^^^^^^^
  File "/home/shreesha/anaconda3/envs/sktime-dev/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py", line 291, in __call__
    return self.fn(*self.args, **self.kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shreesha/anaconda3/envs/sktime-dev/lib/python3.11/site-packages/joblib/parallel.py", line 589, in __call__
    return [func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shreesha/anaconda3/envs/sktime-dev/lib/python3.11/site-packages/joblib/parallel.py", line 589, in 
    return [func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shreesha/sktime_shreesha/sktime/sktime/transformations/series/summarize.py", line 489, in _window_feature
    feat = feat.rolling(window_length).apply(summarizer, raw=True)
           ^^^^
UnboundLocalError: cannot access local variable 'feat' where it is not associated with a value
"""

The above exception was the direct cause of the following exception:
...
--> 754         raise self._result
    755     return self._result
    756 finally:

UnboundLocalError: cannot access local variable 'feat' where it is not associated with a value
```

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

I would like to be guided on what i should be looking at to resolve this issue as this issue does not occur when i run just 
```
import numpy as np
import pandas as pd
from sktime.transformations.series.summarize import WindowSummarizer

df = pd.DataFrame(np.arange(10))
transformer = WindowSummarizer(lag_feature={"mean": [[-1, 1], [-1, 2]]}, n_jobs=1)
df_transformed = transformer.fit_transform(df)

print(df_transformed)
```
 I get the expected output of 0.5 for the 1st row 2nd column as well.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
(https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by fkiraly** [comment]
Hm, superficially, it looks like it is changing the indices in a way that doesn't seem obvious?

FYI @benHeid, @Ram0nB

File: sktime/transformations/series/summarize.py [hunk 1]
```diff
@@ -453,43 +453,37 @@ class description for in-depth explanation.
...
+                    .bfill()
                 )
     else:
-        if bfill is False:
-            feat = Z.shift(lag)
-        else:
-            feat = Z.shift(lag).bfill()
         if isinstance(Z, pd.core.groupby.generic.SeriesGroupBy) and callable(
             summarizer
         ):
```
Comment (benHeid) [inline sktime/transformations/series/summarize.py:489]: Line 490 is the cause of the third error you mentioned. In that case `feat` is not defined yet. Thus, in Line 490 and 494, it must probably be `Z.apply(...`

**Comment by benHeid** [comment]
Regarding the failure with the hierarchical data (first error), could you try to remove lines 328 to 335 in the summarizer file. I suppose that this lines are not needed anymore. At least the first if block is currently duplicating the non-time index levels. 

@fkiraly Do you have any idea why there the index levels are duplicated?

File: sktime/transformations/series/summarize.py (hunk 1 above)
Comment (ShreeshaM07) [reply sktime/transformations/series/summarize.py:489]: Yes the `feat` not being defined yet has now been resolved I will now try to fix the other error .

**Comment by ShreeshaM07** [comment]
> Regarding the failure with the hierarchical data (first error), could you try to remove lines 328 to 335 in the summarizer file. I suppose that this lines are not needed anymore. At least the first if block is currently duplicating the non-time index levels.
> 
> @fkiraly Do you have any idea why there the index levels are duplicated?

I have skipped the lines 329 to 333 and am returning the original `X` DataFrame `isinstance(X.index, pd.MultiIndex)` is `True`.This seems to be solving the issue for some of the files I tested on. It also seems to be avoiding any NaN values.

**Comment by ShreeshaM07** [comment]
In the `test-nosoftdeps-full` it is now failing these tests in the `test_window_summarizer.py`'s `test_windowsummarizer` function for these tests after I removed the lines 329 to 333 in `summarize.py`
```
=========================== short test summary info ============================
FAILED sktime/transformations/series/tests/test_window_summarizer.py::test_windowsummarizer[None-column_names6-y6-None-None] - assert False
 +  where False = all([False, True])
FAILED sktime/transformations/series/tests/test_window_summarizer.py::test_windowsummarizer[kwargs3-column_names3-y3-None-None] - AssertionError: assert 1 == 4
 +  where 1 = len(['y'])
 +  and   4 = len(['y_lag_1', 'y_mean_1_3', 'y_mean_1_12', 'y_std_1_4'])
FAILED sktime/transformations/series/tests/test_window_summarizer.py::test_windowsummarizer[kwargs5-column_names5-y5-None-None] - AssertionError: assert 1 == 4
 +  where 1 = len(['y'])
 +  and   4 = len(['y_lag_1', 'y_mean_1_3', 'y_mean_1_12', 'y_std_1_4'])
FAILED sktime/transformations/series/tests/test_window_summarizer.py::test_windowsummarizer[kwargs4-column_names4-y4-None-None] - AssertionError: assert 1 == 4
 +  where 1 = len(['y'])
 +  and   4 = len(['y_lag_1', 'y_mean_1_3', 'y_mean_1_12', 'y_std_1_4'])
= 4 failed, 15088 passed, 4621 skipped, 7 xfailed, 832 warnings in 968.48s (0:16:08) =
make: *** [Makefile:68: test_softdeps_full] Error 1
Error: Process completed with exit code 2.
```
Could you please help me out as to why these are failing.

File: sktime/transformations/series/summarize.py [hunk 2]
```diff
@@ -323,27 +323,33 @@ def _transform(self, X, y=None):
             bfill = True
         else:
             bfill = False
+        concat_flag = False
         for cols in target_cols:
             if isinstance(X.index, pd.MultiIndex):
```
Comment (benHeid) [inline sktime/transformations/series/summarize.py:328]: I think you should also remove this if statement. Thus the code should like:

```python
...
for cols in target_cols:
                df = Parallel(n_jobs=self.n_jobs)(
                     delayed(_window_feature)(X.loc[:, [cols]], **kwargs, bfill=bfill)
                     for _index, kwargs in func_dict.iterrows()
                 )
                 Xt = pd.concat(df, axis=1)
                 Xt = Xt.add_prefix(str(cols) + "_")
                 Xt_out.append(Xt)
.... 
```

File: sktime/transformations/series/summarize.py [hunk 3]
```diff
@@ -323,27 +323,33 @@ def _transform(self, X, y=None):
...
-        Xt_out_df = pd.concat(Xt_out, axis=1)
-        Xt_return = pd.concat([Xt_out_df, X.drop(target_cols, axis=1)], axis=1)
-
-        Xt_return = Xt_return.loc[idx]
-        return Xt_return
+                Xt = pd.concat(df, axis=1)
+                Xt = Xt.add_prefix(str(cols) + "_")
+                Xt_out.append(Xt)
+                concat_flag = True
+        if concat_flag is True:
```
Comment (benHeid) [inline sktime/transformations/series/summarize.py:345]: The concat flag should not be necessary

File: sktime/transformations/series/summarize.py [hunk 4]
```diff
@@ -323,27 +323,33 @@ def _transform(self, X, y=None):
             bfill = True
         else:
             bfill = False
+        concat_flag = False
```
Comment (benHeid) [inline sktime/transformations/series/summarize.py:326]: Sorry my first guess seems to be a mistake. Thus, I think the changes in this method should be reverted.

File: sktime/transformations/series/summarize.py [hunk 5]
```diff
@@ -453,55 +459,54 @@ class description for in-depth explanation.
         if isinstance(Z, pd.core.groupby.generic.SeriesGroupBy):
```
Comment (benHeid) [inline sktime/transformations/series/summarize.py:459]: I think the issue with the failing tests is located in this if condition. 

A rolling - summarize operation and a SeriesGroupBy object seems to double the index level 0. Not sure why. Thus, the easiest fix would be to drop the first index in 474. However, this is only fixing the symptom and not really the cause... Thus @fkiraly do you have a better idea.

**Comment by ShreeshaM07** [comment]
For the Hierarchical data we must perform shift first then rolling then apply the summarizer and in case of normal series we must perform rolling first then summarizer then shift. The values are now printing as expected . The NaN values in the top rows in heirarchical data is due to the lag values, the summarizer is calculated once we cross the lag value time and then the window_length is reached. This is supposed to be the expected behaviour itself. The current PR now works correctly and has passed all checks. Please let me know if anything else needs to be added.

**Comment by benHeid** [comment]
Could you please add a test for the fix. You can use the code snippet from the issue.

**Comment by ShreeshaM07** [comment]
I have made a new PR with all details about tests addition ,also added myself as a contributor in CONTRIBUTORS.md . Please let me know if any modification is required.

**Pull request closed without merging**